### ISSUE TITLE
[Bug]: tap to focus while zoomed out causes crash

### ISSUE THREAD

**Issue by cdiddy77** [ThibaultBee/StreamPack#97]
### Version

2.5.2

### Environment that reproduces the issue

Samsung Galaxy S22

### RTMP/SRT/... Server

Not streaming

### Audio configuration

N/A

### Video configuration

N/A

### Is it reproducible in the demos application?

Yes

### Reproduction steps

zoom out (zoom ration < 1) 
tap on a region outside of the natural zoom=1 rectangle

### Expected result

to auto-focus on the point tapped on. 

### Actual result

Crash

### Additional context

The proximate cause of the crash is that inside `FocusMetering.startFocusAndMetering` (CameraSettings.kt) the cropRegion returned from` Zoom.getCropRegion()` has an x,y that are negative (bc zoomRatio < 1) that gets passed to `getMeteringRectangles`, which then attempts to create a metering rectangle with negative boundaries. `MeteringRectangle` throws an exception: `"rect.left must be nonnegative"`

**Comment by ThibaultBee** [comment]
Oh nice one.
Could you provide the full exception?

Now I wonder, what will be the best answer to this issue. 
Should we clamp focus rectangle >= 0?
Or should we cancel focus when we detect the issue?
Or somehow synchronize everything?

**Comment by cdiddy77** [comment]
The relevant exception stack trace:
```
"main@15148" prio=5 tid=0x2 nid=NA runnable
  java.lang.Thread.State: RUNNABLE
	  at com.android.internal.util.Preconditions.checkArgumentNonnegative(Preconditions.java:324)
	  at android.hardware.camera2.params.MeteringRectangle.<init>(MeteringRectangle.java:138)
	  at io.github.thibaultbee.streampack.utils.FocusMetering$Companion.getMeteringRect(CameraSettings.kt:982)
	  at io.github.thibaultbee.streampack.utils.FocusMetering$Companion.getMeteringRectangles(CameraSettings.kt:926)
	  at io.github.thibaultbee.streampack.utils.FocusMetering$Companion.access$getMeteringRectangles(CameraSettings.kt:848)
	  at io.github.thibaultbee.streampack.utils.FocusMetering.startFocusAndMetering(CameraSettings.kt:719)
	  at io.github.thibaultbee.streampack.utils.FocusMetering.onTap(CameraSettings.kt:816)
	  at io.github.thibaultbee.streampack.views.PreviewView.performClick(PreviewView.kt:144)
	  at io.github.thibaultbee.streampack.views.PreviewView.onTouchEvent(PreviewView.kt:129)
	  at android.view.View.dispatchTouchEvent(View.java:16212)
```

Great question, I guess in the best of all worlds it would "just work" but that seems difficult. And also, in my specific case on an S22, what's happening under the covers, I believe, is that the camera is automatically switching to the wide-angle lens, which, on my device, is a fixed-field lens and thus, no focusing is possible.

I don't really see a clear best option, but I am certainly no expert. On my device, simply canceling the focus would be fine, since the fixed lens can't focus anyway. In any case you could argue that this is both easy and reasonable. I guess the second choice would be clamping. If it were me, I would cancel and log a message that leads the developer to understand that if they want, they could clamp it themselves.

**Comment by ThibaultBee** [comment]
I tried to reproduce this issue in the demo but I could not. 
Does it happen when you zoom in or is it just when you zoom out? Have you changed something in the code?
I didn't know it was possible to have zoom < 1.

Unfortunately I don't have a device to test this.

Could you add log in the `FocusMetering` class? Could you also log `availableRatioRange` and everything you find usefull?

**Comment by ThibaultBee** [comment]
Could you test https://github.com/ThibaultBee/StreamPack/pull/98?

**Comment by cdiddy77** [comment]
It does not crash when I zoom out, and there is lots of 
```
2024-02-06 14:19:03.044  6781-6781  ViewRootIm...nActivity] io....thibaultbee.streampack.sample  I  ViewPostIme pointer 1
2024-02-06 14:19:03.056  6781-6781  >>TEST                  io....thibaultbee.streampack.sample  E  cropRegion: Rect(0, 0 - 4080, 3060)
2024-02-06 14:19:03.056  6781-6781  >>TEST                  io....thibaultbee.streampack.sample  E  adjustedPoint: PointF(0.6418467, 0.423781)
2024-02-06 14:19:03.056  6781-6781  >>TEST                  io....thibaultbee.streampack.sample  E  size: 0.16666667
2024-02-06 14:19:03.056  6781-6781  >>TEST                  io....thibaultbee.streampack.sample  E  centerX: 2618 - centerY: 1296
2024-02-06 14:19:03.056  6781-6781  >>TEST                  io....thibaultbee.streampack.sample  E  width: 680.0 - height: 510.0
2024-02-06 14:19:03.057  6781-6781  >>TEST                  io....thibaultbee.streampack.sample  E  focusRect1: Rect(2278, 1041 - 2958, 1551)
2024-02-06 14:19:03.057  6781-6781  >>TEST                  io....thibaultbee.streampack.sample  E  focusRect2: Rect(2278, 1041 - 2958, 1551)
2024-02-06 14:19:03.057  6781-6781  >>TEST                  io....thibaultbee.streampack.sample  E  cropRegion: Rect(0, 0 - 4080, 3060)
2024-02-06 14:19:03.057  6781-6781  >>TEST                  io....thibaultbee.streampack.sample  E  adjustedPoint: PointF(0.6418467, 0.423781)
2024-02-06 14:19:03.057  6781-6781  >>TEST                  io....thibaultbee.streampack.sample  E  size: 0.25
2024-02-06 14:19:03.057  6781-6781  >>TEST                  io....thibaultbee.streampack.sample  E  centerX: 2618 - centerY: 1296
2024-02-06 14:19:03.057  6781-6781  >>TEST                  io....thibaultbee.streampack.sample  E  width: 1020.0 - height: 765.0
2024-02-06 14:19:03.058  6781-6781  >>TEST                  io....thibaultbee.streampack.sample  E  focusRect1: Rect(2108, 913 - 3128, 1678)
2024-02-06 14:19:03.058  6781-6781  >>TEST                  io....thibaultbee.streampack.sample  E  focusRect2: Rect(2108, 913 - 3128, 1678)
2024-02-06 14:19:03.206  6781-6781  GestureDetector         io....thibaultbee.streampack.sample  I  handleMessage TAP

```

When I tap outside of the 1x zoom rectangle, I see:
```
2024-02-06 14:21:13.480  6781-6781  ViewRootIm...nActivity] io....thibaultbee.streampack.sample  I  ViewPostIme pointer 1
2024-02-06 14:21:13.493  6781-6781  >>TEST                  io....thibaultbee.streampack.sample  E  cropRegion: Rect(0, 0 - 4080, 3060)
2024-02-06 14:21:13.493  6781-6781  >>TEST                  io....thibaultbee.streampack.sample  E  adjustedPoint: PointF(0.013833943, 0.6899047)
2024-02-06 14:21:13.493  6781-6781  >>TEST                  io....thibaultbee.streampack.sample  E  size: 0.16666667
2024-02-06 14:21:13.493  6781-6781  >>TEST                  io....thibaultbee.streampack.sample  E  centerX: 56 - centerY: 2111
2024-02-06 14:21:13.494  6781-6781  >>TEST                  io....thibaultbee.streampack.sample  E  width: 680.0 - height: 510.0
2024-02-06 14:21:13.494  6781-6781  >>TEST                  io....thibaultbee.streampack.sample  E  focusRect1: Rect(-284, 1856 - 396, 2366)
2024-02-06 14:21:13.494  6781-6781  >>TEST                  io....thibaultbee.streampack.sample  E  focusRect2: Rect(0, 1856 - 396, 2366)
2024-02-06 14:21:13.494  6781-6781  >>TEST                  io....thibaultbee.streampack.sample  E  cropRegion: Rect(0, 0 - 4080, 3060)
2024-02-06 14:21:13.495  6781-6781  >>TEST                  io....thibaultbee.streampack.sample  E  adjustedPoint: PointF(0.013833943, 0.6899047)
2024-02-06 14:21:13.495  6781-6781  >>TEST                  io....thibaultbee.streampack.sample  E  size: 0.25
2024-02-06 14:21:13.495  6781-6781  >>TEST                  io....thibaultbee.streampack.sample  E  centerX: 56 - centerY: 2111
2024-02-06 14:21:13.495  6781-6781  >>TEST                  io....thibaultbee.streampack.sample  E  width: 1020.0 - height: 765.0
2024-02-06 14:21:13.495  6781-6781  >>TEST                  io....thibaultbee.streampack.sample  E  focusRect1: Rect(-454, 1728 - 566, 2493)
2024-02-06 14:21:13.495  6781-6781  >>TEST                  io....thibaultbee.streampack.sample  E  focusRect2: Rect(0, 1728 - 566, 2493)
2024-02-06 14:21:16.767  6781-6781  ViewRootIm...nActivity] io....thibaultbee.streampack.sample  I  onDisplayChanged oldDisplayState=2 newDisplayState=2

```

**Comment by ThibaultBee** [comment]
Logs were unintentional.
I just removed there.

**Comment by ThibaultBee** [comment]
Released in 2.6.0
